cache/cache_reservation_manager.h

```diff
@@ -13,54 +13,91 @@
 #include <cstddef>
 #include <cstdint>
 #include <memory>
+#include <mutex>
 #include <vector>
 
 #include "cache/cache_entry_roles.h"
+#include "cache/cache_key.h"
 #include "rocksdb/cache.h"
 #include "rocksdb/slice.h"
 #include "rocksdb/status.h"
-#include "table/block_based/block_based_table_reader.h"
 #include "util/coding.h"
 
 namespace ROCKSDB_NAMESPACE {
+// CacheReservationManager is an interface for reserving cache space for the
+// memory used
+class CacheReservationManager
+    : public std::enable_shared_from_this<CacheReservationManager> {
```

Comment:
[Note] we need to store pointer to CacheReservationManager in CacheReservationHandle so we can't really do "std::unique_ptr" in call-sites, enforced by compilation error from enable_shared_from_this if do so (hence extra changes included in this PR)

Benchmark will be included for those changes for concerns on the slow-down from shared_ptr.

A relevant discussion https://github.com/facebook/rocksdb/pull/9073/files#r748641039 - I can't really think of a way to improve this so I'll leave it as it is right now but welcome any ideas/thoughts!

Reply:
> Do we need enable_shared_from_this?

This reason to have "enable_shared_from_this" is more for creating a shared_ptr from *this* pointer (https://github.com/facebook/rocksdb/pull/9748/files#diff-b95042a5d09f06814f920b8c334a4f0aff0574c41981d1fc82a93d1c87e774b5R240). Based on a quick search, extending from enable_shared_from_this and calling shared_from_this is the only standard way of doing so.

Reply:
> I can see that this leads to ...

This refers to the use of shared_from_this (i.e., what I assume is the reason you felt the need to extend enable_shared_from_this), not the extension mechanism itself.

Reply:
> I can see that this leads to ...

This refers to the use of shared_from_this (i.e., what I assume is the reason you felt the need to extend enable_shared_from_this), not the extension mechanism itself.

Reply:
> I can see that this leads to ...

This refers to the use of shared_from_this (i.e., what I assume is the reason you felt the need to extend enable_shared_from_this), not the extension mechanism itself.

Reply:
> It's not necessary to change in this PR anyways.

Okay - then I won't make any immediate decision about this in this PR. 

> But it seems to sacrifice some ability to notice reservation leaks, like in cases where we have test coverage that's not specific to CRM.

But for my future reference (tech-debt kind of thing), can you explain more about how in this case, "enable_shared_from_this" seems to sacrifice some ability to notice reservation leaks?

An alternative is to use raw pointer but maybe there is some other design pattern to explore in future for this.

Reply:
> But for my future reference (tech-debt kind of thing), can you explain more about how in this case, "enable_shared_from_this" seems to sacrifice some ability to notice reservation leaks?

Allowing the handle to extend the manager's lifetime means we could leak the handle until the end of the program assuming there's no CRM-specific test coverage for that case (presumably CRM-specific test coverage would assert memory is reduced when it should be). If there's non-CRM test coverage for that code path we would notice such a leak via ASAN because it'd attempt to access a deleted CRM. Of course the risk is there's no test coverage for the leak path in which case extending the manager's lifetime until the handle is finally destroyed is beneficial.